core-rust/engine-state-api-server/src/engine_state_api/factories.rs

```diff
@@ -0,0 +1,210 @@
+            non_instantiated_node_ids: self.staged_node_ids.clone(),
+        }
+    }
+
+    /// Creates a loader of raw data living in an object's Main module.
+    pub fn create_data_loader(
+        &'s self,
+    ) -> EngineStateDataLoader<'s, UnmergeableSubstateDatabaseOverlay<'s, S>> {
+        EngineStateDataLoader {
+            reader: SystemDatabaseReader::new(&self.database),
+        }
+    }
+
+    /// Creates a higher-level loader of data living in an object's Role Assignment module.
+    pub fn create_object_role_assignment_loader(
+        &'s self,
+    ) -> ObjectRoleAssignmentLoader<'s, UnmergeableSubstateDatabaseOverlay<'s, S>> {
+        ObjectRoleAssignmentLoader {
+            meta_loader: self.create_meta_loader(),
+            data_loader: self.create_data_loader(),
+        }
+    }
+
+    /// Creates a higher-level loader of data living in an object's Royalty module.
+    pub fn create_object_royalty_loader(
+        &'s self,
+    ) -> ObjectRoyaltyLoader<'s, UnmergeableSubstateDatabaseOverlay<'s, S>> {
+        ObjectRoyaltyLoader {
+            meta_loader: self.create_meta_loader(),
+            data_loader: self.create_data_loader(),
+        }
+    }
+
+    /// Creates a higher-level loader of data living in an object's Metadata module.
+    pub fn create_object_metadata_loader(
+        &'s self,
+    ) -> ObjectMetadataLoader<'s, UnmergeableSubstateDatabaseOverlay<'s, S>> {
+        ObjectMetadataLoader {
+            loader: self.create_data_loader(),
+        }
+    }
+
+    /// Checks whether the given entity should be automatically instantiated on access, given the
+    /// current database state.
+    fn requires_instantiation(&self, node_id: &NodeId) -> bool {
+        if !node_id.is_global_virtual() {
+            // No matter if it exists or not, we shouldn't instantiate:
+            return false;
+        }
+        // We check whether it is missing by trying to load the `TypeInfo`:
+        return SystemDatabaseReader::new(&self.database)
+            .get_type_info(node_id)
+            .is_err();
+    }
+
+    /// Triggers automatic instantiation of the given entity, using a handcrafted transaction on the
+    /// internally-staged database.
+    fn instantiate(&mut self, node_id: &NodeId) -> CommitResult {
+        let intent = create_intent_forcing_instantiation(node_id);
+
+        // Note: here, we initialize a new `ScryptoVm` instance every time - however, this is
+        // currently a very lightweight operation. At the same time, we would NOT be able to feel
+        // the benefits of caching an instance (since we do not use any WASM here). If these
+        // assumptions change in future, then it will make most sense to turn the `Self` into a
+        // long-lived service holding a `ScryptoVm` as its dependency.
+        let scrypto_vm = ScryptoVm::<DefaultWasmEngine>::default();
+
+        let receipt = execute_and_commit_transaction(
```

Comment:
Who came up with this instantiation through manifest solution lol!

A small side effect of this approach is an update to the existing transaction tracker. Would this affect any of engine state APIs?

Reply:
This approach was decided before https://github.com/radixdlt/babylon-node/pull/978 after some discussion with Josh (see https://radixdlt.atlassian.net/browse/NODE-648?focusedCommentId=20074). Other alternatives would either require either a too-big refactor in the Engine, or a too-error-prone hardcoding of values in the Node, so here we are...

Conceptually, it is the most bulletproof way (i.e. it cannot go wrong, it returns exactly what a real transaction would do). Practically, its runtime cost is acceptable (see https://github.com/radixdlt/babylon-node/pull/978#issuecomment-2194189897).

> A small side effect of this approach is an update to the existing transaction tracker. Would this affect any of engine state APIs?

It definitely cannot affect the behavior of current endpoints.
If we ever add some kind of "batch" endpoint (e.g. read the state of TransactionTracker + read the state of non-instantiated Account in the same call), then we will have to be careful (i.e. isolate those reads; do not re-use the staged store).